iminuit/util.py

```diff
@@ -100,13 +100,12 @@ def arguments_from_funccode(f):
     """
     fc = fc_or_c(f)
     vnames = fc.co_varnames
+    nargs = fc.co_argcount
     # bound method and fake function will be None
-    if is_bound(f):
-        # bound method dock off self
-        return list(vnames[1:fc.co_argcount])
-    else:
-        # unbound and fakefunc
-        return list(vnames[:fc.co_argcount])
+    args = vnames[1 if is_bound(f) else 0:nargs]
+    if not args:
+        raise RuntimeError('Function has variable number of arguments')
```

Comment:
I would put `ValueError`. It's a classical case of a user passing in an invalid value (i.e. a function that can't be processed).

Reply:
Sounds good, but then we should change all the errors. Please have a look into the file, all the functions there throw RuntimeError. I followed that lead instead of throwing a more specific error. In fact, the error type does not matter, because any error thrown in one of these utility functions is converted into a TypeError in `better_arg_spec`.

Reply:
I wasn't aware. Then let's keep as-is.

Reply:
Perhaps we can improve the code in utils.py at some later point. Raising errors, then capturing them to print a message and then throwing another error again looks a bit inelegant.

Reply:
Catching a RunTimeError is terrible. It can probably occur from, well, runtime issues.
It should never be caught and masked.

The pattern to catch errors is not bad, but for those cases, a custom exception that's internal to the library should be created.

Reply:
The raised errors here are used as an internal control structure. I think that's not a good use for errors. I would not catch errors internally to transform them and give other errors to the user, that seems superfluous. Raising and catching errors is very slow in C++, probably less so in Python (because everything is slow in Python), but it should not be used for an internal control structure. I would either hand the original error to the user or not raise errors but indicate internal failures in some other way, e.g. by returning None. Errors are there to propagate to the user.

Reply:
Relevent SO discussion: https://stackoverflow.com/questions/7274310/python-using-exceptions-for-control-flow-considered-bad

Reply:
So we agree.